File: packages/cli/src/commands/projects.ts
import { Command } from "commander";
import chalk from "chalk";
import ora from "ora";
import { resolve } from "node:path";
import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { createApiClient } from "../lib/api-client.js";
import { findProjectConfig } from "../lib/config.js";
import { success, info, keyValue, warn } from "../lib/output.js";
import { handleError } from "../lib/shared.js";

/* ─── remb projects ─── */

export const projectsCommand = new Command("projects")
  .description("Manage projects — list, switch active project")
  .addCommand(projectsListCommand())
  .addCommand(projectsUseCommand());

/* ─── remb projects list ─── */

/* ─── remb projects use ─── */

function projectsUseCommand() {
  return new Command("use")
    .alias("switch")
    .description("Set the active project for this workspace — writes/updates .remb.yml")
    .argument("<slug>", "Project slug to activate")
    .option("--api-url <url>", "API server URL to write into .remb.yml")
    .addHelpText(
      "after",
      `\nExamples:\n  $ remb projects use my-app\n  $ remb projects switch my-app\n  $ remb projects use my-app --api-url https://remb.vercel.app`,
    )
    .action(async (slug: string, opts) => {
      const cwd = process.cwd();
      const existing = findProjectConfig(cwd);
      const configPath = resolve(existing?.dir ?? cwd, ".remb.yml");
      const apiUrl = opts.apiUrl ?? existing?.config.api_url ?? "http://localhost:3000";
      const ide = existing?.config.ide;

      // Validate the slug exists remotely
      const spinner = ora(`Looking up project "${slug}"...`).start();
      try {
        const client = createApiClient();
        const { projects } = await client.listProjects({ status: "active", limit: 200 });
        const found = projects.find((p) => p.slug === slug);
        spinner.stop();
        if (!found) {
          const available = projects.map((p) => `  ${chalk.cyan(p.slug)} — ${p.name}`).join("\n");
          warn(`Project "${slug}" not found. Your projects:\n${available}`);
          process.exit(1);
        }

        // Write .remb.yml
        const lines = [
          "# Remb project configuration",
          `# Updated by remb projects use`,
          "",
          `project: ${slug}`,
          `api_url: ${apiUrl}`,
        ];
        if (ide) lines.push(`ide: ${ide}`);
        lines.push("");
        writeFileSync(configPath, lines.join("\n"), "utf-8");

        success(`Active project set to ${chalk.bold(found.name)} (${chalk.cyan(slug)})`);
        keyValue("Config", configPath);
        keyValue("Features", String(found.feature_count));
        keyValue("Entries", String(found.entry_count));
        if (found.repo_name) keyValue("Repo", found.repo_name);
      } catch (err) {
        spinner.stop();
        handleError(err);
      }
    });
}

/* ─── remb projects list ─── */

function projectsListCommand() {
  return new Command("list")
    .alias("ls")
    .description("List all projects")
    .option("--status <status>", "Filter by status")
    .option("-l, --limit <n>", "Max results", "50")
    .option("--format <format>", "Output format: table, json, markdown", "table")
    .addHelpText(
      "after",
      `\nExamples:\n  $ remb projects list\n  $ remb projects ls --format json\n  $ remb projects list --status active`,
    )
    .action(async (opts) => {
      const spinner = ora("Fetching projects...").start();
      try {
        const client = createApiClient();
        const { projects, total } = await client.listProjects({
          status: opts.status,
          limit: parseInt(opts.limit, 10),
        });
        spinner.stop();

        if (projects.length === 0) {
          info("No projects found. Create one with: remb init");
          return;
        }

        info(`${total} project${total === 1 ? "" : "s"} found\n`);

        if (opts.format === "json") {
          console.log(JSON.stringify(projects, null, 2));
          return;
        }

        if (opts.format === "markdown") {
          for (const p of projects) {
            console.log(`### ${p.name}`);
            console.log(`- **Slug**: ${p.slug} | **Status**: ${p.status}`);
            console.log(`- **Language**: ${p.language ?? "—"} | **Branch**: ${p.branch}`);
            console.log(`- **Features**: ${p.feature_count} | **Entries**: ${p.entry_count}`);
            if (p.description) console.log(`- **Description**: ${p.description}`);
            if (p.repo_url) console.log(`- **Repo**: ${p.repo_url}`);
            console.log();
          }
          return;
        }

        // Table format
        const statusColors: Record<string, (s: string) => string> = {
          active: chalk.green,
          archived: chalk.dim,
          draft: chalk.yellow,
        };

        for (const p of projects) {
          const statusFn = statusColors[p.status] ?? chalk.white;
          const lang = p.language ? chalk.dim(`[${p.language}]`) : "";
          console.log(
            `${statusFn(p.status.padEnd(10))} ${chalk.bold(p.name.slice(0, 30).padEnd(32))} ${chalk.dim(p.slug.padEnd(25))} ${lang}`
          );
          console.log(
            `           ${chalk.dim(`${p.feature_count} features, ${p.entry_count} entries`)}${p.repo_name ? chalk.dim(` · ${p.repo_name}`) : ""}`
          );
        }
      } catch (err) {
        spinner.stop();
        handleError(err);
      }
    });
}

File: packages/cli-go/cmd/projects.go
package cmd

import (
	"encoding/json"
	"fmt"
	"os"
	"text/tabwriter"

	"github.com/richie/remb/internal/api"
	"github.com/richie/remb/internal/output"
	"github.com/spf13/cobra"
)

var projectsCmd = &cobra.Command{
	Use:     "projects",
	Aliases: []string{"proj"},
	Short:   "Manage projects — list and inspect",
}

// ─── remb projects list ────────────────────────────────────────────────

var projListStatus string
var projListLimit int
var projListFormat string

var projectsListCmd = &cobra.Command{
	Use:     "list",
	Aliases: []string{"ls"},
	Short:   "List all projects",
	RunE:    runProjectsList,
}

func runProjectsList(cmd *cobra.Command, args []string) error {
	fmt.Print("⠋ Fetching projects...")

	client, err := api.NewClient()
	if err != nil {
		fmt.Print("\r\033[K")
		output.Error(err.Error())
		os.Exit(1)
	}

	params := map[string]string{}
	if projListStatus != "" {
		params["status"] = projListStatus
	}
	if projListLimit > 0 {
		params["limit"] = fmt.Sprintf("%d", projListLimit)
	}

	resp, err := client.ListProjects(params)
	fmt.Print("\r\033[K")
	if err != nil {
		output.Error(err.Error())
		os.Exit(1)
	}

	if len(resp.Projects) == 0 {
		output.Info("No projects found. Create one with: remb init")
		return nil
	}

	output.Info(fmt.Sprintf("%d project%s found\n", resp.Total, plural(resp.Total)))

	switch projListFormat {
	case "json":
		enc := json.NewEncoder(os.Stdout)
		enc.SetIndent("", "  ")
		return enc.Encode(resp.Projects)
	case "markdown":
		for _, p := range resp.Projects {
			fmt.Printf("### %s\n", p.Name)
			fmt.Printf("- **Slug**: %s | **Status**: %s\n", p.Slug, p.Status)
			fmt.Printf("- **Language**: %s | **Branch**: %s\n", derefOr(p.Language, "—"), p.Branch)
			fmt.Printf("- **Features**: %d | **Entries**: %d\n", p.FeatureCount, p.EntryCount)
			if p.Description != nil {
				fmt.Printf("- **Description**: %s\n", *p.Description)
			}
			if p.RepoURL != nil {
				fmt.Printf("- **Repo**: %s\n", *p.RepoURL)
			}
			fmt.Println()
		}
		return nil
	default:
		w := tabwriter.NewWriter(os.Stdout, 0, 4, 2, ' ', 0)
		fmt.Fprintf(w, "STATUS\tNAME\tSLUG\tLANG\tFEATURES\tENTRIES\n")
		for _, p := range resp.Projects {
			fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%d\t%d\n",
				p.Status,
				truncate(p.Name, 30),
				p.Slug,
				derefOr(p.Language, "—"),
				p.FeatureCount,
				p.EntryCount,
			)
		}
		return w.Flush()
	}
}

func plural(n int) string {
	if n == 1 {
		return ""
	}
	return "s"
}

func derefOr(s *string, fallback string) string {
	if s != nil && *s != "" {
		return *s
	}
	return fallback
}

func truncate(s string, max int) string {
	if len(s) <= max {
		return s
	}
	return s[:max-1] + "…"
}

func init() {
	projectsListCmd.Flags().StringVar(&projListStatus, "status", "", "Filter by status")
	projectsListCmd.Flags().IntVarP(&projListLimit, "limit", "l", 50, "Max results")
	projectsListCmd.Flags().StringVar(&projListFormat, "format", "table", "Output format: table, json, markdown")

	projectsCmd.AddCommand(projectsListCmd)
	rootCmd.AddCommand(projectsCmd)
}
